devito/types/tensor.py

```diff
@@ -31,16 +67,22 @@ class TensorFunction(AbstractTensor, Differentiable):
         Staggering of each component, needs to have the size of the tensor. Defaults
         to the Dimensions.
     """
+    _is_TimeDependent = False
     _sub_type = Function
-    _op_priority = Differentiable._op_priority + 2.
     _class_priority = 10
+    _op_priority = Differentiable._op_priority + 1.
 
     def __init_finalize__(self, *args, **kwargs):
-        self._is_symmetric = kwargs.get('symmetric', True)
-        self._is_diagonal = kwargs.get('diagonal', False)
-        self._staggered = kwargs.get('staggered', self.space_dimensions)
-        self._grid = kwargs.get('grid')
-        self._space_order = kwargs.get('space_order', 1)
+        if args:
+            comps = flatten(args[2])
+            grid = comps[0].grid
+            dimensions = None if grid else comps[0].dimensions
```

Comment:
you mean the other way round? if yes (maybe I'm wrong...) is this path actually tested? if not, what would `comps[0]` be to carry `dimensions` ? another tensorial object? 

to ease understanding of this class, you may consider adding some asserts and/or checks on the input values

Reply:
>what would comps[0] be to carry dimensions
A differentiable expression for example, i.e `div(VectorFunction)` will be a Tensor with derivatives as component (or evaluated derivatives) or any expression carrying dimensions but no grid.